libraries/AP_AccelCal/AccelCalibrator.cpp

```diff
@@ -0,0 +1,477 @@
+        set_status(ACCEL_CAL_FAILED);
+    }
+}
+
+// returns spherical fit paramteters
+void AccelCalibrator::get_calibration(Vector3f& offset) const {
+    offset = -_param_struct.offset;
+}
+
+// returns axis aligned ellipsoidal fit parameters
+void AccelCalibrator::get_calibration(Vector3f& offset, Vector3f& diag) const {
+    offset = -_param_struct.offset;
+    diag = _param_struct.diag;
+}
+
+// returns generic ellipsoidal fit parameters
+void AccelCalibrator::get_calibration(Vector3f& offset, Vector3f& diag, Vector3f& offdiag) const {
+    offset = -_param_struct.offset;
+    diag = _param_struct.diag;
+    offdiag = _param_struct.offdiag;
+}
+
+/////////////////////////////////////////////////////////////
+////////////////////// PRIVATE METHODS //////////////////////
+/////////////////////////////////////////////////////////////
+
+/*
+  The sample acceptance distance is determined as follows:
+  For any regular polyhedron with triangular faces, the angle theta subtended
+  by two closest points is defined as
+ 
+       theta = arccos(cos(A)/(1-cos(A)))
+ 
+  Where:
+       A = (4pi/F + pi)/3
+  and
+       F = 2V - 4 is the number of faces for the polyhedron in consideration,
+       which depends on the number of vertices V
+ 
+  The above equation was proved after solving for spherical triangular excess
+  and related equations.
+ */
+bool AccelCalibrator::accept_sample(const Vector3f& sample)
+{
+    static const uint16_t faces = 2*_conf_num_samples-4;
+    static const float a = (4.0f * M_PI_F / (3.0f * faces)) + M_PI_F / 3.0f;
+    static const float theta = 0.5f * acosf(cosf(a) / (1.0f - cosf(a)));
+
+    if (_sample_buffer == NULL) {
+        return false;
+    }
+
+    float min_distance = GRAVITY_MSS * 2*sinf(theta/2);
+
+    for(uint8_t i=0; i < _samples_collected; i++) {
+        Vector3f other_sample;
+        get_sample(i, other_sample);
+
+        if ((other_sample - sample).length() < min_distance){
+            return false;
+        }
+    }
+    return true;
+}
+
+// sets status of calibrator and takes appropriate actions
+void AccelCalibrator::set_status(enum accel_cal_status_t status) {
+    switch (status) {
+        case ACCEL_CAL_NOT_STARTED:                 
+            //Calibrator not started
+            _status = ACCEL_CAL_NOT_STARTED;
+
+            _samples_collected = 0;
+            if (_sample_buffer != NULL) {
+                free(_sample_buffer);
+                _sample_buffer = NULL;
+            }
+
+            break;
+
+        case ACCEL_CAL_WAITING_FOR_ORIENTATION:     
+            //Callibrator has been started and is waiting for user to ack after orientation setting
+            if (!running()) {
+                _samples_collected = 0;
+                if (_sample_buffer == NULL) {
+                    _sample_buffer = (struct AccelSample*)malloc(sizeof(struct AccelSample)*_conf_num_samples);
```

Comment:
just use calloc instead of malloc and memset
or use new[], as we zero all memory anyway